### PR TITLE
DPL: add metric for shm memory usage

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@davidrohr as discussed.

File: Framework/Core/include/Framework/DataProcessingStats.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ struct DataProcessingStats {
   std::atomic<int> totalProcessedSize = 0;
   std::atomic<int> totalSigusr1 = 0;
   std::atomic<int> consumedTimeframes = 0;
+  std::atomic<int> availableManagedShm = 0;
```
Comment (davidrohr) [inline Framework/Core/include/Framework/DataProcessingStats.h]: Is this in bytes or in GB? If it is bytes, the `int` will overflow

File: Framework/Core/src/CommonServices.cxx [hunk 2]
```diff
@@ -472,6 +479,7 @@ auto sendRelayerMetrics(ServiceRegistry& registry, DataProcessingStats& stats) -
   monitoring.send(Metric{(stats.lastProcessedSize / (stats.lastLatency.maxLatency ? stats.lastLatency.maxLatency : 1) / 1000), "input_rate_mb_s"}
                     .addTag(Key::Subsystem, Value::DPL));
   monitoring.send(Metric{((float)performedComputationsSinceLastUpdate / (float)timeSinceLastUpdate) * 1000, "processing_rate_hz"}.addTag(Key::Subsystem, Value::DPL));
+  monitoring.send(Metric{(uint64_t)stats.availableManagedShm, "available_managed_shm"}.addTag(Key::Subsystem, Value::DPL));
```
Comment (davidrohr) [inline Framework/Core/src/CommonServices.cxx]: could we appent `_[segment_id]` here, in the end the monitoring will receive both, and we need to distinguish.

**Comment by davidrohr** [comment]
And then it actually doesn't work, I get an error like
```
[21:29:29][ERROR] Could not find management segment for shmid '17a7d816'
```
followed by an exception
I think the wrong shmid is computed:
```
[O2PDPSuite/latest] ~/alitest/tmp-small $> ls /dev/shm/
com.freerdp.xfreerdp.tsmf_000000000000923F  fmq_eca859e3_cv  fmq_eca859e3_m_0  fmq_eca859e3_mng  sem.fmq_eca859e3_mtx
```

**Comment by davidrohr** [comment]
Ah, and one last point, can we enable it only for the readout-proxy, or enable it via command line option / env variable?
Just to reduce the metric load a bit.

File: Framework/Core/src/CommonServices.cxx (hunk 2 above)
Comment (ktf) [reply Framework/Core/src/CommonServices.cxx]: Shouldn't we distinguish based on the dataprocessor_id global tag?

**Comment by ktf** [comment]
What did you use as session id?

File: Framework/Core/src/CommonServices.cxx (hunk 2 above)
Comment (davidrohr) [reply Framework/Core/src/CommonServices.cxx]: When we have 2 NUMA domains, with 2 DPL chains, don't we have each dataprocessor_id twice? Only this id we put in the XML will be unique.

**Comment by davidrohr** [comment]
> What did you use as session id?

`--session default`. However, I think DDS somehow overwrite the session via the odc plugin. It is always dífferent, even though I always provide `--session default`. Can't we get the session that is used from FairMQ?

**Comment by ktf** [comment]
If the property is set, sure. This might also affect memory cleanup...

**Comment by davidrohr** [comment]
> If the property is set, sure. This might also affect memory cleanup...

OK, then we should try to get it from there. And perhaps we should put the whole complex in a `try` / `catch` to avoid that an exception kills us.

**Comment by ktf** [comment]
Ok. I have updated this. I tested it and it still works correctly without DDS.

File: Framework/Core/src/CommonServices.cxx [hunk 3]
```diff
@@ -438,6 +440,17 @@ auto sendRelayerMetrics(ServiceRegistry& registry, DataProcessingStats& stats) -
...
   }
+  // Derive the amount of shared memory used
+  auto& runningWorkflow = registry.get<RunningWorkflowInfo const>();
+  using namespace fair::mq::shmem;
+  auto& spec = registry.get<DeviceSpec const>();
+
+  // FIXME: Ugly, but we do it only every 5 seconds...
+  if (spec.name == "readout-proxy") {
+    auto device = registry.get<RawDeviceService>().device();
+    stats.availableManagedShm.store(Monitor::GetFreeMemory(SessionId{device->fConfig->GetProperty<std::string>("session")}, runningWorkflow.shmSegmentId));
```
Comment (davidrohr) [inline Framework/Core/src/CommonServices.cxx]: Now it works perfectly. But can you put a try {...} catch {} around this. I want to make sure we don't create a build tonight that will crash on the EPN due to exception.

**Comment by davidrohr** [comment]
tested on the EPN, merging

**Pull request merged**